File: dotnet-desktop-guide/winforms/forms/snippets/events/vb/AsyncEventHandlers.vb
Imports System
Imports System.Net.Http
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows.Forms

Partial Public Class ExampleForm
    Inherits Form

    Private WithEvents downloadButton As Button
    Private WithEvents processButton As Button
    Private WithEvents complexButton As Button
    Private WithEvents badButton As Button
    Private WithEvents legacyButton As Button
    Private resultTextBox As TextBox
    Private statusLabel As Label
    Private progressBar As ProgressBar

    Public Sub New()
        Size = New System.Drawing.Size(400, 600)
        downloadButton = New Button With {.Text = "Download Data"}
        processButton = New Button With {.Text = "Process Data"}
        complexButton = New Button With {.Text = "Complex Operation"}
        badButton = New Button With {.Text = "Bad Example (Deadlock)"}
        resultTextBox = New TextBox With {.Multiline = True, .Width = 300, .Height = 200}
        statusLabel = New Label With {.Text = "Status: Ready"}
        progressBar = New ProgressBar With {.Width = 300}

        'Arrange controls
        Dim Margin = 20
        Dim spacing = 10

        'Buttons stacked vertically.
        downloadButton.Location = New System.Drawing.Point(Margin, Margin)
        downloadButton.Width = 300

        processButton.Location = New System.Drawing.Point(Margin, downloadButton.Bottom + spacing)
        processButton.Width = downloadButton.Width

        complexButton.Location = New System.Drawing.Point(Margin, processButton.Bottom + spacing)
        complexButton.Width = downloadButton.Width

        badButton.Location = New System.Drawing.Point(Margin, complexButton.Bottom + spacing)
        badButton.Width = downloadButton.Width

        'Status And progress below buttons.
        statusLabel.AutoSize = True
        statusLabel.Location = New System.Drawing.Point(Margin, badButton.Bottom + spacing)

        progressBar.Location = New System.Drawing.Point(Margin, statusLabel.Bottom + spacing)
        progressBar.Width = downloadButton.Width

        'Result text box below progress bar.
        resultTextBox.Location = New System.Drawing.Point(Margin, progressBar.Bottom + spacing)

        Controls.Add(downloadButton)
        Controls.Add(processButton)
        Controls.Add(complexButton)
        Controls.Add(badButton)
        Controls.Add(resultTextBox)
        Controls.Add(statusLabel)
        Controls.Add(progressBar)
    End Sub

    ' <snippet_BasicAsyncEventHandler>
    Private Async Sub downloadButton_Click(sender As Object, e As EventArgs) Handles downloadButton.Click
        downloadButton.Enabled = False
        statusLabel.Text = "Downloading..."

        Try
            Using httpClient As New HttpClient()
                Dim content As String = Await httpClient.GetStringAsync("https://github.com/dotnet/docs/raw/refs/heads/main/README.md")

                ' Update UI with the result
                resultTextBox.Text = content
                statusLabel.Text = "Download complete"
            End Using
        Catch ex As Exception
            statusLabel.Text = $"Error: {ex.Message}"
        Finally
            downloadButton.Enabled = True
        End Try
    End Sub
    ' </snippet_BasicAsyncEventHandler>

    ' <snippet_DeadlockAntiPattern>
    ' DON'T DO THIS - causes deadlocks
    Private Sub badButton_Click(sender As Object, e As EventArgs) Handles badButton.Click
        Try
            ' This blocks the UI thread and causes a deadlock
            Dim content As String = DownloadPageContentAsync().GetAwaiter().GetResult()
            resultTextBox.Text = content
        Catch ex As Exception
            MessageBox.Show($"Error: {ex.Message}")
        End Try
    End Sub

    Private Async Function DownloadPageContentAsync() As Task(Of String)
        Using httpClient As New HttpClient()
            Return Await httpClient.GetStringAsync("https://github.com/dotnet/docs/raw/refs/heads/main/README.md")
        End Using
    End Function
    ' </snippet_DeadlockAntiPattern>

    ' <snippet_InvokeAsyncNet9>
    Private Async Sub processButton_Click(sender As Object, e As EventArgs) Handles processButton.Click
        processButton.Enabled = False

        ' Start background work
        Await Task.Run(Async Function()
                           For i As Integer = 0 To 100 Step 10
                               ' Simulate work
                               Await Task.Delay(200)

                               ' Create local variable to avoid closure issues
                               Dim currentProgress As Integer = i

                               ' Update UI safely from background thread
                               Await progressBar.InvokeAsync(Sub()
                                                                 progressBar.Value = currentProgress
                                                                 statusLabel.Text = $"Progress: {currentProgress}%"
                                                             End Sub)
                           Next
                       End Function)

        processButton.Enabled = True
    End Sub
    ' </snippet_InvokeAsyncNet9>

    ' <snippet_InvokeAsyncUIThread>
    Private Async Sub complexButton_Click(sender As Object, e As EventArgs) Handles complexButton.Click
        ' For VB.NET, we use a simpler approach since async lambdas with CancellationToken are more complex
        Await Me.InvokeAsync(Sub()
                                 ' This runs on UI thread
                                 statusLabel.Text = "Starting complex operation..."
                             End Sub)

        ' Perform the async operation
        Dim result As String = Await SomeAsyncApiCall()

        ' Update UI after completion
        Await Me.InvokeAsync(Sub()
                                 resultTextBox.Text = result
                                 statusLabel.Text = "Operation completed"
                             End Sub)
    End Sub

    Private Async Function SomeAsyncApiCall() As Task(Of String)
        Using httpClient As New HttpClient()
            Return Await httpClient.GetStringAsync("https://github.com/dotnet/docs/raw/refs/heads/main/README.md")
        End Using
    End Function
    ' </snippet_InvokeAsyncUIThread>

    ' <snippet_LegacyNetFramework>
    Private Async Sub legacyButton_Click(sender As Object, e As EventArgs) Handles legacyButton.Click
        Dim legacyButton As Button = TryCast(sender, Button)
        legacyButton.Enabled = False

        Try
            ' Move to background thread to avoid blocking UI
            Await Task.Run(Async Function()
                               Dim result As String = Await SomeAsyncOperation()

                               ' Marshal back to UI thread
                               Me.Invoke(New Action(Sub()
                                                        resultTextBox.Text = result
                                                        statusLabel.Text = "Complete"
                                                    End Sub))
                           End Function)
        Finally
            ' This runs on the UI thread since the await completed
            legacyButton.Enabled = True
        End Try
    End Sub

    Private Async Function SomeAsyncOperation() As Task(Of String)
        Using httpClient As New HttpClient()
            Return Await httpClient.GetStringAsync("https://github.com/dotnet/docs/raw/refs/heads/main/README.md")
        End Using
    End Function
    ' </snippet_LegacyNetFramework>
End Class

File: dotnet-desktop-guide/winforms/forms/snippets/events/cs/AsyncEventHandlers.cs
using System;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace AsyncEventHandlers;

public partial class ExampleForm : Form
{
    private Button downloadButton;
    private Button processButton;
    private Button complexButton;
    private Button badButton;
    private TextBox resultTextBox;
    private Label statusLabel;
    private ProgressBar progressBar;

    public ExampleForm()
    {
        Size = new System.Drawing.Size(400, 600);
        downloadButton = new Button { Text = "Download Data" };
        processButton = new Button { Text = "Process Data" };
        complexButton = new Button { Text = "Complex Operation" };
        badButton = new Button { Text = "Bad Example (Deadlock)" };
        resultTextBox = new TextBox { Multiline = true, Width = 300, Height = 200 };
        statusLabel = new Label { Text = "Status: Ready" };
        progressBar = new ProgressBar { Width = 300 };
        downloadButton.Click += downloadButton_Click;
        processButton.Click += processButton_Click;
        complexButton.Click += complexButton_Click;
        badButton.Click += badButton_Click;

        // Arrange controls
        int margin = 20;
        int spacing = 10;

        // Buttons stacked vertically.
        downloadButton.Location = new System.Drawing.Point(margin, margin);
        downloadButton.Width = 300;

        processButton.Location = new System.Drawing.Point(margin, downloadButton.Bottom + spacing);
        processButton.Width = downloadButton.Width;

        complexButton.Location = new System.Drawing.Point(margin, processButton.Bottom + spacing);
        complexButton.Width = downloadButton.Width;

        badButton.Location = new System.Drawing.Point(margin, complexButton.Bottom + spacing);
        badButton.Width = downloadButton.Width;

        // Status and progress below buttons.
        statusLabel.AutoSize = true;
        statusLabel.Location = new System.Drawing.Point(margin, badButton.Bottom + spacing);

        progressBar.Location = new System.Drawing.Point(margin, statusLabel.Bottom + spacing);
        progressBar.Width = downloadButton.Width;

        // Result text box below progress bar.
        resultTextBox.Location = new System.Drawing.Point(margin, progressBar.Bottom + spacing);


        Controls.Add(downloadButton);
        Controls.Add(processButton);
        Controls.Add(complexButton);
        Controls.Add(badButton);
        Controls.Add(resultTextBox);
        Controls.Add(statusLabel);
        Controls.Add(progressBar);
    }

    // <snippet_BasicAsyncEventHandler>
    private async void downloadButton_Click(object sender, EventArgs e)
    {
        downloadButton.Enabled = false;
        statusLabel.Text = "Downloading...";
        
        try
        {
            using var httpClient = new HttpClient();
            string content = await httpClient.GetStringAsync("https://github.com/dotnet/docs/raw/refs/heads/main/README.md");
            
            // Update UI with the result
            resultTextBox.Text = content;
            statusLabel.Text = "Download complete";
        }
        catch (Exception ex)
        {
            statusLabel.Text = $"Error: {ex.Message}";
        }
        finally
        {
            downloadButton.Enabled = true;
        }
    }
    // </snippet_BasicAsyncEventHandler>

    // <snippet_DeadlockAntiPattern>
    // DON'T DO THIS - causes deadlocks
    private void badButton_Click(object sender, EventArgs e)
    {
        try
        {
            // This blocks the UI thread and causes a deadlock
            string content = DownloadPageContentAsync().GetAwaiter().GetResult();
            resultTextBox.Text = content;
        }
        catch (Exception ex)
        {
            MessageBox.Show($"Error: {ex.Message}");
        }
    }

    private async Task<string> DownloadPageContentAsync()
    {
        using var httpClient = new HttpClient();
        await Task.Delay(2000); // Simulate delay
        return await httpClient.GetStringAsync("https://github.com/dotnet/docs/raw/refs/heads/main/README.md");
    }
    // </snippet_DeadlockAntiPattern>

    // <snippet_InvokeAsyncNet9>
    private async void processButton_Click(object sender, EventArgs e)
    {
        processButton.Enabled = false;
        
        // Start background work
        await Task.Run(async () =>
        {
            for (int i = 0; i <= 100; i += 10)
            {
                // Simulate work
                await Task.Delay(200);
                
                // Create local variable to avoid closure issues
                int currentProgress = i;
                
                // Update UI safely from background thread
                await progressBar.InvokeAsync(() =>
                {
                    progressBar.Value = currentProgress;
                    statusLabel.Text = $"Progress: {currentProgress}%";
                });
            }
        });
        
        processButton.Enabled = true;
    }
    // </snippet_InvokeAsyncNet9>

    // <snippet_InvokeAsyncUIThread>
    private async void complexButton_Click(object sender, EventArgs e)
    {
        await this.InvokeAsync(async (cancellationToken) =>
        {
            // This runs on UI thread but doesn't block it
            statusLabel.Text = "Starting complex operation...";
            
            var result = await SomeAsyncApiCall();
            
            // Update UI directly since we're already on UI thread
            resultTextBox.Text = result;
            statusLabel.Text = "Operation completed";
        });
    }

    private async Task<string> SomeAsyncApiCall()
    {
        using var httpClient = new HttpClient();
        return await httpClient.GetStringAsync("https://github.com/dotnet/docs/raw/refs/heads/main/README.md");
    }
    // </snippet_InvokeAsyncUIThread>

    // <snippet_LegacyNetFramework>
    private async void legacyButton_Click(object sender, EventArgs e)
    {
        var legacyButton = sender as Button;
        legacyButton.Enabled = false;
        
        try
        {
            // Move to background thread to avoid blocking UI
            await Task.Run(async () =>
            {
                var result = await SomeAsyncOperation();
                
                // Marshal back to UI thread
                this.Invoke(new Action(() =>
                {
                    resultTextBox.Text = result;
                    statusLabel.Text = "Complete";
                }));
            });
        }
        finally
        {
            // This runs on the UI thread since the await completed
            legacyButton.Enabled = true;
        }
    }

    private async Task<string> SomeAsyncOperation()
    {
        using var httpClient = new HttpClient();
        await Task.Delay(2000); // Simulate delay
        return await httpClient.GetStringAsync("https://github.com/dotnet/docs/raw/refs/heads/main/README.md");
    }
    // </snippet_LegacyNetFramework>
}
